File: udp_server.go
package main

import (
	"fmt"
	"net"
	"sync"
	"time"

	"github.com/google/uuid"
	"github.com/sirupsen/logrus"
)

type UDPClient struct {
	ID           uuid.UUID
	Addr         net.Addr
	Player       *Player
	LastSeen     time.Time
	Sequence     uint32
	AckSequence  uint32
	PendingAcks  map[uint32]*PendingPacket
	SessionID    *int64
	mu           sync.RWMutex
}

type PendingPacket struct {
	Packet    *UDPPacket
	Timestamp time.Time
}

func NewUDPClient(id uuid.UUID, addr net.Addr, name string, sessionID *int64) *UDPClient {
	player := NewPlayer(id, name)
	return &UDPClient{
		ID:          id,
		Addr:        addr,
		Player:      player,
		LastSeen:    time.Now(),
		Sequence:    0,
		AckSequence: 0,
		PendingAcks: make(map[uint32]*PendingPacket),
		SessionID:   sessionID,
	}
}

func (uc *UDPClient) UpdatePosition(x, y float32) {
	uc.mu.Lock()
	defer uc.mu.Unlock()
	uc.Player.X = x
	uc.Player.Y = y
	uc.LastSeen = time.Now()
}

func (uc *UDPClient) UpdateHealth(health float32) {
	uc.mu.Lock()
	defer uc.mu.Unlock()
	uc.Player.Health = health
}

func (uc *UDPClient) AddScore(points uint32) {
	uc.mu.Lock()
	defer uc.mu.Unlock()
	uc.Player.Score += points
}

func (uc *UDPClient) NextSequence() uint32 {
	uc.mu.Lock()
	defer uc.mu.Unlock()
	uc.Sequence++
	return uc.Sequence
}

func (uc *UDPClient) IsTimeout() bool {
	uc.mu.RLock()
	defer uc.mu.RUnlock()
	return time.Since(uc.LastSeen) > 30*time.Second
}

func (uc *UDPClient) AddPendingAck(packet *UDPPacket) {
	uc.mu.Lock()
	defer uc.mu.Unlock()
	uc.PendingAcks[packet.Sequence] = &PendingPacket{
		Packet:    packet,
		Timestamp: time.Now(),
	}
}

func (uc *UDPClient) RemovePendingAck(sequence uint32) bool {
	uc.mu.Lock()
	defer uc.mu.Unlock()
	_, exists := uc.PendingAcks[sequence]
	if exists {
		delete(uc.PendingAcks, sequence)
	}
	return exists
}

func (uc *UDPClient) GetTimeoutPackets() []uint32 {
	uc.mu.RLock()
	defer uc.mu.RUnlock()
	
	var timeoutSeqs []uint32
	for seq, pending := range uc.PendingAcks {
		if time.Since(pending.Timestamp) > 100*time.Millisecond {
			timeoutSeqs = append(timeoutSeqs, seq)
		}
	}
	return timeoutSeqs
}

type UDPGameServer struct {
	conn        *net.UDPConn
	clients     map[string]*UDPClient // key: addr.String()
	clientByID  map[uuid.UUID]string  // key: client ID, value: addr.String()
	database    *Database
	mu          sync.RWMutex
}

func NewUDPGameServer(addr string, database *Database) (*UDPGameServer, error) {
	udpAddr, err := net.ResolveUDPAddr("udp", addr)
	if err != nil {
		return nil, fmt.Errorf("failed to resolve UDP address: %w", err)
	}

	conn, err := net.ListenUDP("udp", udpAddr)
	if err != nil {
		return nil, fmt.Errorf("failed to listen on UDP: %w", err)
	}

	logrus.Infof("UDP Game server listening on: %s", addr)

	server := &UDPGameServer{
		conn:       conn,
		clients:    make(map[string]*UDPClient),
		clientByID: make(map[uuid.UUID]string),
		database:   database,
	}

	// Start background tasks
	go server.startHeartbeatTask()
	go server.startCleanupTask()
	go server.startReliabilityTask()

	return server, nil
}

func (ugs *UDPGameServer) Run() error {
	buf := make([]byte, 1500) // MTU size

	for {
		n, addr, err := ugs.conn.ReadFromUDP(buf)
		if err != nil {
			logrus.Errorf("UDP recv error: %v", err)
			continue
		}

		data := buf[:n]
		packet, err := DeserializeUDPPacket(data)
		if err != nil {
			logrus.Warnf("Failed to deserialize packet from %s", addr)
			continue
		}

		go ugs.handlePacket(addr, packet)
	}
}

func (ugs *UDPGameServer) handlePacket(addr *net.UDPAddr, packet *UDPPacket) {
	switch packet.Message.Type {
	case "Heartbeat":
		if data, ok := packet.Message.Data.(map[string]interface{}); ok {
			if playerIDStr, ok := data["player_id"].(string); ok {
				if playerID, err := uuid.Parse(playerIDStr); err == nil {
					if sequence, ok := data["sequence"].(float64); ok {
						ugs.handleHeartbeat(addr, playerID, uint32(sequence))
					}
				}
			}
		}
	case "Ack":
		if data, ok := packet.Message.Data.(map[string]interface{}); ok {
			if sequence, ok := data["sequence"].(float64); ok {
				ugs.handleAck(addr, uint32(sequence))
			}
		}
	case "PlayerMove":
		if data, ok := packet.Message.Data.(map[string]interface{}); ok {
			if playerIDStr, ok := data["player_id"].(string); ok {
				if playerID, err := uuid.Parse(playerIDStr); err == nil {
					if x, ok := data["x"].(float64); ok {
						if y, ok := data["y"].(float64); ok {
							ugs.handlePlayerMove(addr, playerID, float32(x), float32(y), packet.Sequence)
						}
					}
				}
			}
		}
	case "PlayerAction":
		if data, ok := packet.Message.Data.(map[string]interface{}); ok {
			if playerIDStr, ok := data["player_id"].(string); ok {
				if playerID, err := uuid.Parse(playerIDStr); err == nil {
					if action, ok := data["action"].(string); ok {
						ugs.handlePlayerAction(addr, playerID, action, data["data"], packet.Sequence)
					}
				}
			}
		}
	case "Chat":
		if data, ok := packet.Message.Data.(map[string]interface{}); ok {
			if playerIDStr, ok := data["player_id"].(string); ok {
				if playerID, err := uuid.Parse(playerIDStr); err == nil {
					if message, ok := data["message"].(string); ok {
						ugs.handleChat(addr, playerID, message, packet.Sequence)
					}
				}
			}
		}
	}
}

func (ugs *UDPGameServer) handleHeartbeat(addr *net.UDPAddr, playerID uuid.UUID, sequence uint32) {
	ugs.mu.Lock()
	defer ugs.mu.Unlock()

	addrStr := addr.String()

	// Check if this is a new client
	if _, exists := ugs.clients[addrStr]; !exists {
		clientName := fmt.Sprintf("Player_%s", playerID.String()[:8])

		// Create session in database
		var sessionID *int64
		ipStr := addr.IP.String()
		if id, err := ugs.database.CreateSession(playerID, "udp", &ipStr); err != nil {
			logrus.Errorf("Failed to create UDP session: %v", err)
			sessionID = nil
		} else {
			sessionID = &id
		}

		client := NewUDPClient(playerID, addr, clientName, sessionID)

		// Save player to database
		if err := ugs.database.CreateOrUpdatePlayer(client.Player); err != nil {
			logrus.Errorf("Failed to save UDP player to database: %v", err)
		}

		// Log join event
		joinMsg := NewPlayerJoinMessage(playerID, clientName)
		if err := ugs.database.LogEvent(playerID, sessionID, "join", &joinMsg); err != nil {
			logrus.Errorf("Failed to log UDP join event: %v", err)
		}

		ugs.clients[addrStr] = client
		ugs.clientByID[playerID] = addrStr

		logrus.Infof("New UDP client connected: %s (%s) with session %v", clientName, addr, sessionID)

		// Send join message to all clients
		ugs.broadcastReliable(&joinMsg, &addrStr)

		// Send current game state to new client
		ugs.sendGameStateToClient(addr)
	} else {
		// Update last seen for existing client
		if client, exists := ugs.clients[addrStr]; exists {
			client.mu.Lock()
			client.LastSeen = time.Now()
			client.AckSequence = sequence
			client.mu.Unlock()
		}
	}

	// Send ACK
	ugs.sendAck(addr, sequence)
}

func (ugs *UDPGameServer) handleAck(addr *net.UDPAddr, sequence uint32) {
	ugs.mu.RLock()
	client, exists := ugs.clients[addr.String()]
	ugs.mu.RUnlock()

	if exists {
		client.RemovePendingAck(sequence)
	}
}

func (ugs *UDPGameServer) handlePlayerMove(addr *net.UDPAddr, playerID uuid.UUID, x, y float32, sequence uint32) {
	ugs.mu.RLock()
	client, exists := ugs.clients[addr.String()]
	ugs.mu.RUnlock()

	if exists && client.ID == playerID {
		client.UpdatePosition(x, y)

		// Update position in database
		if err := ugs.database.UpdatePlayerPosition(playerID, x, y); err != nil {
			logrus.Errorf("Failed to update UDP player position in database: %v", err)
		}

		// Log move event (less frequent for UDP to avoid spam)
		if sequence%10 == 0 {
			moveMsg := NewPlayerMoveMessage(playerID, x, y)
			if err := ugs.database.LogEvent(playerID, client.SessionID, "move", &moveMsg); err != nil {
				logrus.Errorf("Failed to log UDP move event: %v", err)
			}
		}

		// Send ACK
		ugs.sendAck(addr, sequence)

		// Broadcast move to other clients (unreliable for performance)
		moveMessage := NewPlayerMoveMessage(playerID, x, y)
		addrStr := addr.String()
		ugs.broadcastUnreliable(&moveMessage, &addrStr)
	}
}

func (ugs *UDPGameServer) handlePlayerAction(addr *net.UDPAddr, playerID uuid.UUID, action string, data interface{}, sequence uint32) {
	ugs.mu.RLock()
	client, exists := ugs.clients[addr.String()]
	ugs.mu.RUnlock()

	if exists && client.ID == playerID {
		switch action {
		case "attack":
			logrus.Infof("Player %s performed attack", playerID)

			// Log attack event
			if err := ugs.database.LogEvent(playerID, client.SessionID, "attack", nil); err != nil {
				logrus.Errorf("Failed to log UDP attack event: %v", err)
			}

		case "pickup":
			client.AddScore(10)
			newScore := client.Player.Score
			logrus.Infof("Player %s picked up item, score: %d", playerID, newScore)

			// Update score in database
			if err := ugs.database.UpdatePlayerScore(playerID, newScore); err != nil {
				logrus.Errorf("Failed to update UDP player score in database: %v", err)
			}

			// Log pickup event
			if err := ugs.database.LogEvent(playerID, client.SessionID, "pickup", nil); err != nil {
				logrus.Errorf("Failed to log UDP pickup event: %v", err)
			}

		default:
			logrus.Infof("Unknown action: %s from player %s", action, playerID)
		}

		// Send ACK
		ugs.sendAck(addr, sequence)
	}
}

func (ugs *UDPGameServer) handleChat(addr *net.UDPAddr, playerID uuid.UUID, message string, sequence uint32) {
	ugs.mu.RLock()
	client, exists := ugs.clients[addr.String()]
	ugs.mu.RUnlock()

	if exists && client.ID == playerID {
		// Save chat message to database
		if err := ugs.database.SaveChatMessage(playerID, client.SessionID, message); err != nil {
			logrus.Errorf("Failed to save UDP chat message to database: %v", err)
		}

		// Log chat event
		chatMsg := NewChatMessage(playerID, message)
		if err := ugs.database.LogEvent(playerID, client.SessionID, "chat", &chatMsg); err != nil {
			logrus.Errorf("Failed to log UDP chat event: %v", err)
		}

		// Send ACK
		ugs.sendAck(addr, sequence)

		// Broadcast chat message (reliable)
		addrStr := addr.String()
		ugs.broadcastReliable(&chatMsg, &addrStr)
	}
}

func (ugs *UDPGameServer) sendAck(addr *net.UDPAddr, sequence uint32) {
	ackMessage := NewAckMessage(sequence)
	packet := NewUDPPacket(0, ackMessage, false)
	data, _ := packet.Serialize()

	if _, err := ugs.conn.WriteToUDP(data, addr); err != nil {
		logrus.Errorf("Failed to send ACK to %s: %v", addr, err)
	}
}

func (ugs *UDPGameServer) broadcastReliable(message *GameMessage, exclude *string) {
	ugs.mu.RLock()
	defer ugs.mu.RUnlock()

	for addrStr, client := range ugs.clients {
		if exclude == nil || *exclude != addrStr {
			sequence := client.NextSequence()
			packet := NewUDPPacket(sequence, *message, true)
			client.AddPendingAck(packet)

			data, _ := packet.Serialize()
			if udpAddr, err := net.ResolveUDPAddr("udp", addrStr); err == nil {
				if _, err := ugs.conn.WriteToUDP(data, udpAddr); err != nil {
					logrus.Errorf("Failed to send reliable message to %s: %v", addrStr, err)
				}
			}
		}
	}
}

func (ugs *UDPGameServer) broadcastUnreliable(message *GameMessage, exclude *string) {
	ugs.mu.RLock()
	defer ugs.mu.RUnlock()

	for addrStr := range ugs.clients {
		if exclude == nil || *exclude != addrStr {
			packet := NewUDPPacket(0, *message, false)
			data, _ := packet.Serialize()

			if udpAddr, err := net.ResolveUDPAddr("udp", addrStr); err == nil {
				if _, err := ugs.conn.WriteToUDP(data, udpAddr); err != nil {
					logrus.Errorf("Failed to send unreliable message to %s: %v", addrStr, err)
				}
			}
		}
	}
}

func (ugs *UDPGameServer) sendGameStateToClient(addr *net.UDPAddr) {
	ugs.mu.RLock()
	defer ugs.mu.RUnlock()

	var players []Player
	for _, client := range ugs.clients {
		players = append(players, *client.Player)
	}

	gameStateMessage := NewGameStateMessage(players)
	addrStr := addr.String()

	if client, exists := ugs.clients[addrStr]; exists {
		sequence := client.NextSequence()
		packet := NewUDPPacket(sequence, gameStateMessage, true)
		client.AddPendingAck(packet)

		data, _ := packet.Serialize()
		if _, err := ugs.conn.WriteToUDP(data, addr); err != nil {
			logrus.Errorf("Failed to send game state to %s: %v", addr, err)
		}
	}
}

func (ugs *UDPGameServer) startHeartbeatTask() {
	ticker := time.NewTicker(5 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			ugs.mu.RLock()
			for addrStr, client := range ugs.clients {
				heartbeat := NewHeartbeatMessage(client.ID, 0)
				packet := NewUDPPacket(0, heartbeat, false)
				data, _ := packet.Serialize()

				if udpAddr, err := net.ResolveUDPAddr("udp", addrStr); err == nil {
					if _, err := ugs.conn.WriteToUDP(data, udpAddr); err != nil {
						logrus.Errorf("Failed to send heartbeat to %s: %v", addrStr, err)
					}
				}
			}
			ugs.mu.RUnlock()
		}
	}
}

func (ugs *UDPGameServer) startCleanupTask() {
	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			ugs.mu.Lock()
			var toRemove []string
			var clientIDs []uuid.UUID

			// Check for timed out clients
			for addrStr, client := range ugs.clients {
				if client.IsTimeout() {
					toRemove = append(toRemove, addrStr)
					clientIDs = append(clientIDs, client.ID)
				}
			}

			// Remove timed out clients
			for i, addrStr := range toRemove {
				clientID := clientIDs[i]
				delete(ugs.clients, addrStr)
				delete(ugs.clientByID, clientID)
				logrus.Infof("Removed timed out UDP client: %s (%s)", clientID, addrStr)
			}
			ugs.mu.Unlock()
		}
	}
}

func (ugs *UDPGameServer) startReliabilityTask() {
	ticker := time.NewTicker(50 * time.Millisecond)
	defer ticker.Stop()

	for {
		select {
		case <-ticker.C:
			ugs.mu.RLock()
			for addrStr, client := range ugs.clients {
				timeoutSeqs := client.GetTimeoutPackets()

				for _, sequence := range timeoutSeqs {
					client.mu.RLock()
					if pending, exists := client.PendingAcks[sequence]; exists {
						data, _ := pending.Packet.Serialize()
						client.mu.RUnlock()

						if udpAddr, err := net.ResolveUDPAddr("udp", addrStr); err == nil {
							if _, err := ugs.conn.WriteToUDP(data, udpAddr); err != nil {
								logrus.Errorf("Failed to resend packet %d to %s: %v", sequence, addrStr, err)
							} else {
								// Update timestamp for next timeout check
								client.mu.Lock()
								if pending, exists := client.PendingAcks[sequence]; exists {
									pending.Timestamp = time.Now()
								}
								client.mu.Unlock()
							}
						}
					} else {
						client.mu.RUnlock()
					}
				}
			}
			ugs.mu.RUnlock()
		}
	}
}

func (ugs *UDPGameServer) GetClientCount() int {
	ugs.mu.RLock()
	defer ugs.mu.RUnlock()
	return len(ugs.clients)
}

File: src/udp_server.rs
use std::net::SocketAddr;
use std::sync::Arc;
use tokio::net::UdpSocket;
use tokio::sync::RwLock;
use dashmap::DashMap;
use uuid::Uuid;
use tracing::{info, error, warn};
use std::collections::HashMap;
use tokio::time::{interval, Duration, Instant};

use crate::message::{GameMessage, UdpPacket, Player};
use crate::database::Database;

#[derive(Debug, Clone)]
pub struct UdpClient {
    pub id: Uuid,
    pub addr: SocketAddr,
    pub player: Player,
    pub last_seen: Instant,
    pub sequence: u32,
    pub ack_sequence: u32,
    pub pending_acks: HashMap<u32, (UdpPacket, Instant)>,
    pub session_id: Option<i64>,
}

impl UdpClient {
    pub fn new(id: Uuid, addr: SocketAddr, name: String, session_id: Option<i64>) -> Self {
        let player = Player::new(id, name);
        Self {
            id,
            addr,
            player,
            last_seen: Instant::now(),
            sequence: 0,
            ack_sequence: 0,
            pending_acks: HashMap::new(),
            session_id,
        }
    }

    pub fn update_position(&mut self, x: f32, y: f32) {
        self.player.x = x;
        self.player.y = y;
        self.last_seen = Instant::now();
    }

    pub fn update_health(&mut self, health: f32) {
        self.player.health = health;
    }

    pub fn add_score(&mut self, points: u32) {
        self.player.score += points;
    }

    pub fn next_sequence(&mut self) -> u32 {
        self.sequence += 1;
        self.sequence
    }

    pub fn is_timeout(&self) -> bool {
        self.last_seen.elapsed() > Duration::from_secs(30)
    }

    pub fn add_pending_ack(&mut self, packet: UdpPacket) {
        self.pending_acks.insert(packet.sequence, (packet, Instant::now()));
    }

    pub fn remove_pending_ack(&mut self, sequence: u32) -> bool {
        self.pending_acks.remove(&sequence).is_some()
    }

    pub fn get_timeout_packets(&self) -> Vec<u32> {
        self.pending_acks
            .iter()
            .filter(|(_, (_, timestamp))| timestamp.elapsed() > Duration::from_millis(100))
            .map(|(seq, _)| *seq)
            .collect()
    }
}

pub struct UdpGameServer {
    socket: Arc<UdpSocket>,
    clients: Arc<DashMap<SocketAddr, Arc<RwLock<UdpClient>>>>,
    client_by_id: Arc<DashMap<Uuid, SocketAddr>>,
    database: Database,
}

impl UdpGameServer {
    pub async fn new(addr: &str, database: Database) -> anyhow::Result<Self> {
        let socket = UdpSocket::bind(addr).await?;
        info!("UDP Game server listening on: {}", addr);

        let server = Self {
            socket: Arc::new(socket),
            clients: Arc::new(DashMap::new()),
            client_by_id: Arc::new(DashMap::new()),
            database,
        };

        // Start background tasks
        server.start_heartbeat_task().await;
        server.start_cleanup_task().await;
        server.start_reliability_task().await;

        Ok(server)
    }

    pub async fn run(&self) -> anyhow::Result<()> {
        let mut buf = vec![0u8; 1500]; // MTU size

        loop {
            match self.socket.recv_from(&mut buf).await {
                Ok((size, addr)) => {
                    let data = &buf[..size];
                    if let Ok(packet) = UdpPacket::deserialize(data) {
                        self.handle_packet(addr, packet).await;
                    } else {
                        warn!("Failed to deserialize packet from {}", addr);
                    }
                }
                Err(e) => {
                    error!("UDP recv error: {}", e);
                }
            }
        }
    }

    async fn handle_packet(&self, addr: SocketAddr, packet: UdpPacket) {
        match &packet.message {
            GameMessage::Heartbeat { player_id, sequence } => {
                self.handle_heartbeat(addr, *player_id, *sequence).await;
            }
            GameMessage::Ack { sequence } => {
                self.handle_ack(addr, *sequence).await;
            }
            GameMessage::PlayerMove { player_id, x, y } => {
                self.handle_player_move(addr, *player_id, *x, *y, packet.sequence).await;
            }
            GameMessage::PlayerAction { player_id, action, data } => {
                self.handle_player_action(addr, *player_id, action, data, packet.sequence).await;
            }
            GameMessage::Chat { player_id, message } => {
                self.handle_chat(addr, *player_id, message, packet.sequence).await;
            }
            _ => {}
        }
    }

    async fn handle_heartbeat(&self, addr: SocketAddr, player_id: Uuid, sequence: u32) {
        // Check if this is a new client
        if !self.clients.contains_key(&addr) {
            let client_name = format!("Player_{}", &player_id.to_string()[..8]);
            
            // Create session in database
            let session_id = match self.database.create_session(&player_id, "udp", Some(&addr.ip().to_string())).await {
                Ok(id) => Some(id),
                Err(e) => {
                    error!("Failed to create UDP session: {}", e);
                    None
                }
            };
            
            let mut client = UdpClient::new(player_id, addr, client_name.clone(), session_id);
            
            // Save player to database
            if let Err(e) = self.database.create_or_update_player(&client.player).await {
                error!("Failed to save UDP player to database: {}", e);
            }
            
            // Log join event
            if let Err(e) = self.database.log_event(&player_id, session_id, "join", None).await {
                error!("Failed to log UDP join event: {}", e);
            }
            
            self.clients.insert(addr, Arc::new(RwLock::new(client)));
            self.client_by_id.insert(player_id, addr);
            
            info!("New UDP client connected: {} ({}) with session {:?}", client_name, addr, session_id);
            
            // Send join message to all clients
            let join_message = GameMessage::PlayerJoin {
                player_id,
                name: client_name,
            };
            self.broadcast_reliable(&join_message, Some(addr)).await;
            
            // Send current game state to new client
            self.send_game_state_to_client(addr).await;
        } else {
            // Update last seen for existing client
            if let Some(client_ref) = self.clients.get(&addr) {
                let mut client = client_ref.write().await;
                client.last_seen = Instant::now();
                client.ack_sequence = sequence;
            }
        }
        
        // Send ACK
        self.send_ack(addr, sequence).await;
    }

    async fn handle_ack(&self, addr: SocketAddr, sequence: u32) {
        if let Some(client_ref) = self.clients.get(&addr) {
            let mut client = client_ref.write().await;
            client.remove_pending_ack(sequence);
        }
    }

    async fn handle_player_move(&self, addr: SocketAddr, player_id: Uuid, x: f32, y: f32, sequence: u32) {
        if let Some(client_ref) = self.clients.get(&addr) {
            let mut client = client_ref.write().await;
            if client.id == player_id {
                client.update_position(x, y);
                let session_id = client.session_id;
                drop(client);
                
                // Update position in database
                if let Err(e) = self.database.update_player_position(&player_id, x, y).await {
                    error!("Failed to update UDP player position in database: {}", e);
                }
                
                // Log move event (less frequent for UDP to avoid spam)
                // Only log every 10th move to reduce database load
                if sequence % 10 == 0 {
                    let move_msg = GameMessage::PlayerMove { player_id, x, y };
                    if let Err(e) = self.database.log_event(&player_id, session_id, "move", Some(&move_msg)).await {
                        error!("Failed to log UDP move event: {}", e);
                    }
                }
                
                // Send ACK for reliable message
                self.send_ack(addr, sequence).await;
                
                // Broadcast move to other clients (unreliable for performance)
                let move_message = GameMessage::PlayerMove { player_id, x, y };
                self.broadcast_unreliable(&move_message, Some(addr)).await;
            }
        }
    }

    async fn handle_player_action(&self, addr: SocketAddr, player_id: Uuid, action: &str, _data: &serde_json::Value, sequence: u32) {
        if let Some(client_ref) = self.clients.get(&addr) {
            let mut client = client_ref.write().await;
            if client.id == player_id {
                let session_id = client.session_id;
                
                match action {
                    "attack" => {
                        info!("Player {} performed attack", player_id);
                        
                        // Log attack event
                        if let Err(e) = self.database.log_event(&player_id, session_id, "attack", None).await {
                            error!("Failed to log UDP attack event: {}", e);
                        }
                    }
                    "pickup" => {
                        client.add_score(10);
                        let new_score = client.player.score;
                        info!("Player {} picked up item, score: {}", player_id, new_score);
                        drop(client);
                        
                        // Update score in database
                        if let Err(e) = self.database.update_player_score(&player_id, new_score).await {
                            error!("Failed to update UDP player score in database: {}", e);
                        }
                        
                        // Log pickup event
                        if let Err(e) = self.database.log_event(&player_id, session_id, "pickup", None).await {
                            error!("Failed to log UDP pickup event: {}", e);
                        }
                    }
                    _ => {
                        info!("Unknown action: {} from player {}", action, player_id);
                    }
                }
                
                // Send ACK for reliable message
                self.send_ack(addr, sequence).await;
            }
        }
    }

    async fn handle_chat(&self, addr: SocketAddr, player_id: Uuid, message: &str, sequence: u32) {
        if let Some(client_ref) = self.clients.get(&addr) {
            let client = client_ref.read().await;
            if client.id == player_id {
                let session_id = client.session_id;
                
                // Save chat message to database
                if let Err(e) = self.database.save_chat_message(&player_id, session_id, message).await {
                    error!("Failed to save UDP chat message to database: {}", e);
                }
                
                // Log chat event
                let chat_msg = GameMessage::Chat {
                    player_id,
                    message: message.to_string(),
                };
                if let Err(e) = self.database.log_event(&player_id, session_id, "chat", Some(&chat_msg)).await {
                    error!("Failed to log UDP chat event: {}", e);
                }
                
                // Send ACK
                self.send_ack(addr, sequence).await;
                drop(client);
                
                // Broadcast chat message (reliable)
                self.broadcast_reliable(&chat_msg, Some(addr)).await;
            }
        }
    }

    async fn send_ack(&self, addr: SocketAddr, sequence: u32) {
        let ack_message = GameMessage::Ack { sequence };
        let packet = UdpPacket::new(0, ack_message, false);
        let data = packet.serialize();
        
        if let Err(e) = self.socket.send_to(&data, addr).await {
            error!("Failed to send ACK to {}: {}", addr, e);
        }
    }

    async fn broadcast_reliable(&self, message: &GameMessage, exclude: Option<SocketAddr>) {
        for client_ref in self.clients.iter() {
            let client_addr = *client_ref.key();
            if exclude.map_or(true, |addr| addr != client_addr) {
                let mut client = client_ref.value().write().await;
                let sequence = client.next_sequence();
                let packet = UdpPacket::new(sequence, message.clone(), true);
                client.add_pending_ack(packet.clone());
                
                let data = packet.serialize();
                if let Err(e) = self.socket.send_to(&data, client_addr).await {
                    error!("Failed to send reliable message to {}: {}", client_addr, e);
                }
            }
        }
    }

    async fn broadcast_unreliable(&self, message: &GameMessage, exclude: Option<SocketAddr>) {
        for client_ref in self.clients.iter() {
            let client_addr = *client_ref.key();
            if exclude.map_or(true, |addr| addr != client_addr) {
                let packet = UdpPacket::new(0, message.clone(), false);
                let data = packet.serialize();
                
                if let Err(e) = self.socket.send_to(&data, client_addr).await {
                    error!("Failed to send unreliable message to {}: {}", client_addr, e);
                }
            }
        }
    }

    async fn send_game_state_to_client(&self, addr: SocketAddr) {
        let mut players = Vec::new();
        for client_ref in self.clients.iter() {
            let client = client_ref.value().read().await;
            players.push(client.player.clone());
        }

        let game_state_message = GameMessage::GameState {
            players,
            timestamp: std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .unwrap()
                .as_secs(),
        };

        if let Some(client_ref) = self.clients.get(&addr) {
            let mut client = client_ref.write().await;
            let sequence = client.next_sequence();
            let packet = UdpPacket::new(sequence, game_state_message, true);
            client.add_pending_ack(packet.clone());
            
            let data = packet.serialize();
            if let Err(e) = self.socket.send_to(&data, addr).await {
                error!("Failed to send game state to {}: {}", addr, e);
            }
        }
    }

    async fn start_heartbeat_task(&self) {
        let clients = self.clients.clone();
        let socket = self.socket.clone();
        
        tokio::spawn(async move {
            let mut interval = interval(Duration::from_secs(5));
            
            loop {
                interval.tick().await;
                
                // Send heartbeat requests to all clients
                for client_ref in clients.iter() {
                    let client_addr = *client_ref.key();
                    let client = client_ref.value().read().await;
                    
                    let heartbeat = GameMessage::Heartbeat {
                        player_id: client.id,
                        sequence: 0,
                    };
                    let packet = UdpPacket::new(0, heartbeat, false);
                    let data = packet.serialize();
                    
                    if let Err(e) = socket.send_to(&data, client_addr).await {
                        error!("Failed to send heartbeat to {}: {}", client_addr, e);
                    }
                }
            }
        });
    }

    async fn start_cleanup_task(&self) {
        let clients = self.clients.clone();
        let client_by_id = self.client_by_id.clone();
        
        tokio::spawn(async move {
            let mut interval = interval(Duration::from_secs(10));
            
            loop {
                interval.tick().await;
                
                let mut to_remove = Vec::new();
                
                // Check for timed out clients
                for client_ref in clients.iter() {
                    let client_addr = *client_ref.key();
                    let client = client_ref.value().read().await;
                    
                    if client.is_timeout() {
                        to_remove.push((client_addr, client.id));
                    }
                }
                
                // Remove timed out clients
                for (addr, client_id) in to_remove {
                    // Get session_id before removing client
                    let session_id = if let Some(client_ref) = clients.get(&addr) {
                        client_ref.read().await.session_id
                    } else {
                        None
                    };
                    
                    clients.remove(&addr);
                    client_by_id.remove(&client_id);
                    info!("Removed timed out UDP client: {} ({})", client_id, addr);
                    
                    // Note: In a real implementation, you'd use a channel to communicate with the main loop
                    // to handle session cleanup and leave message broadcasting
                }
            }
        });
    }

    async fn start_reliability_task(&self) {
        let clients = self.clients.clone();
        let socket = self.socket.clone();
        
        tokio::spawn(async move {
            let mut interval = interval(Duration::from_millis(50)); // Check every 50ms
            
            loop {
                interval.tick().await;
                
                // Resend timed out reliable messages
                for client_ref in clients.iter() {
                    let client_addr = *client_ref.key();
                    let mut client = client_ref.value().write().await;
                    
                    let timeout_sequences = client.get_timeout_packets();
                    
                    for sequence in timeout_sequences {
                        if let Some((packet, _)) = client.pending_acks.get(&sequence).cloned() {
                            let data = packet.serialize();
                            if let Err(e) = socket.send_to(&data, client_addr).await {
                                error!("Failed to resend packet {} to {}: {}", sequence, client_addr, e);
                            } else {
                                // Update timestamp for next timeout check
                                client.pending_acks.insert(sequence, (packet, Instant::now()));
                            }
                        }
                    }
                }
            }
        });
    }

    pub fn get_client_count(&self) -> usize {
        self.clients.len()
    }
}